pkg/kubectl/resource_printer.go

```diff
@@ -35,66 +35,94 @@ import (
+// generic is true).
+func GetPrinter(format, formatArgument string) (ResourcePrinter, bool, error) {
 	var printer ResourcePrinter
 	switch format {
 	case "json":
-		printer = &JSONPrinter{version, convertor}
+		printer = &JSONPrinter{}
 	case "yaml":
-		printer = &YAMLPrinter{version, convertor}
+		printer = &YAMLPrinter{}
 	case "template":
 		if len(formatArgument) == 0 {
-			return nil, fmt.Errorf("template format specified but no template given")
+			return nil, false, fmt.Errorf("template format specified but no template given")
 		}
 		var err error
-		printer, err = NewTemplatePrinter([]byte(formatArgument), version, convertor)
+		printer, err = NewTemplatePrinter([]byte(formatArgument))
 		if err != nil {
-			return nil, fmt.Errorf("error parsing template %s, %v\n", formatArgument, err)
+			return nil, false, fmt.Errorf("error parsing template %s, %v\n", formatArgument, err)
 		}
 	case "templatefile":
 		if len(formatArgument) == 0 {
-			return nil, fmt.Errorf("templatefile format specified but no template file given")
+			return nil, false, fmt.Errorf("templatefile format specified but no template file given")
 		}
 		data, err := ioutil.ReadFile(formatArgument)
 		if err != nil {
-			return nil, fmt.Errorf("error reading template %s, %v\n", formatArgument, err)
+			return nil, false, fmt.Errorf("error reading template %s, %v\n", formatArgument, err)
 		}
-		printer, err = NewTemplatePrinter(data, version, convertor)
+		printer, err = NewTemplatePrinter(data)
 		if err != nil {
-			return nil, fmt.Errorf("error parsing template %s, %v\n", string(data), err)
+			return nil, false, fmt.Errorf("error parsing template %s, %v\n", string(data), err)
 		}
 	case "":
-		printer = defaultPrinter
+		return nil, false, nil
 	default:
-		return nil, fmt.Errorf("output format %q not recognized", format)
+		return nil, false, fmt.Errorf("output format %q not recognized", format)
 	}
-	return printer, nil
+	return printer, true, nil
 }
 
 // ResourcePrinter is an interface that knows how to print runtime objects.
 type ResourcePrinter interface {
-	// Print receives an arbitrary object, formats it and prints it to a writer.
+	// Print receives a runtime object, formats it and prints it to a writer.
 	PrintObj(runtime.Object, io.Writer) error
 }
 
-// JSONPrinter is an implementation of ResourcePrinter which outputs an object as JSON.
-// The input object is assumed to be in the internal version of an API and is converted
-// to the given version first.
-type JSONPrinter struct {
-	version   string
+// ResourcePrinterFunc is a function that can print objects
+type ResourcePrinterFunc func(runtime.Object, io.Writer) error
+
+// PrintObj implements ResourcePrinter
+func (fn ResourcePrinterFunc) PrintObj(obj runtime.Object, w io.Writer) error {
+	return fn(obj, w)
+}
+
+// VersionedPrinter takes runtime objects and ensures they are converted to a given API version
+// prior to being passed to a nested printer.
+type VersionedPrinter struct {
```

Comment:
We should put that example into test-cmd.sh then.  :)

-o json should be outputting a minionlist - is it?


> On Jan 13, 2015, at 5:56 PM, Daniel Smith <notifications@github.com> wrote:
> 
> In pkg/kubectl/resource_printer.go:
> 
> > -// JSONPrinter is an implementation of ResourcePrinter which outputs an object as JSON.
> > -// The input object is assumed to be in the internal version of an API and is converted
> > -// to the given version first.
> > -type JSONPrinter struct {
> > -	version   string
> > +// ResourcePrinterFunc is a function that can print objects
> > +type ResourcePrinterFunc func(runtime.Object, io.Writer) error
> > +
> > +// PrintObj implements ResourcePrinter
> > +func (fn ResourcePrinterFunc) PrintObj(obj runtime.Object, w io.Writer) error {
> > +	return fn(obj, w)
> > +}
> > +
> > +// VersionedPrinter takes runtime objects and ensures they are converted to a given API version
> > +// prior to being passed to a nested printer.
> > +type VersionedPrinter struct {
> They were when I last touched this code! In fact the last thing I did to this code was to fix this very problem...
> 
> —
> Reply to this email directly or view it on GitHub.
>